tfe/resource_tfe_workspace_run.go

```diff
@@ -0,0 +1,157 @@
+	tfe.RunPending:        true,
+	tfe.RunPlanQueued:     true,
+	tfe.RunPlanning:       true,
+	tfe.RunCostEstimating: true,
+	tfe.RunPolicyChecking: true,
+	tfe.RunQueuing:        true,
+	tfe.RunFetching:       true,
+}
+
+var planTerminalStatuses = map[tfe.RunStatus]bool{
+	tfe.RunPlanned:            true,
+	tfe.RunPlannedAndFinished: true,
+	tfe.RunErrored:            true,
+	tfe.RunCostEstimated:      true,
+	tfe.RunPolicyChecked:      true,
+	tfe.RunPolicySoftFailed:   true,
+	tfe.RunPolicyOverride:     true,
+}
+
+var applyPendingStatuses = map[tfe.RunStatus]bool{
+	tfe.RunConfirmed:   true,
+	tfe.RunApplyQueued: true,
+	tfe.RunApplying:    true,
+	tfe.RunQueuing:     true,
+	tfe.RunFetching:    true,
+}
+
+var applyDoneStatuses = map[tfe.RunStatus]bool{
+	tfe.RunApplied: true,
+	tfe.RunErrored: true,
+}
+
+var confirmationDoneStatuses = map[tfe.RunStatus]bool{
+	tfe.RunConfirmed:   true,
+	tfe.RunApplyQueued: true,
+	tfe.RunApplying:    true,
+}
+
+func resourceTFEWorkspaceRun() *schema.Resource {
+	return &schema.Resource{
+		Create:        resourceTFEWorkspaceRunCreate,
+		Delete:        resourceTFEWorkspaceRunDelete,
+		Read:          resourceTFEWorkspaceRunRead,
+		Update:        resourceTFEWorkspaceRunUpdate,
+		SchemaVersion: 1,
+		Schema: map[string]*schema.Schema{
+			"workspace_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"apply": {
+				Type:         schema.TypeList,
+				Elem:         resourceTFEWorkspaceRunSchema(),
+				Optional:     true,
+				AtLeastOneOf: []string{"apply", "destroy"},
+				MaxItems:     1,
+			},
+			"destroy": {
+				Type:     schema.TypeList,
+				Elem:     resourceTFEWorkspaceRunSchema(),
+				Optional: true,
+				MaxItems: 1,
+			},
+		},
+	}
+}
+
+func resourceTFEWorkspaceRunCreate(d *schema.ResourceData, meta interface{}) error {
+	// var isDestroyRun & currentRetryAttempts is declared for the sole purpose of code readability
+	isDestroyRun := false
+	currentRetryAttempts := 0
+	return createWorkspaceRun(d, meta, isDestroyRun, currentRetryAttempts)
+}
+
+func resourceTFEWorkspaceRunDelete(d *schema.ResourceData, meta interface{}) error {
+	// var isDestroyRun & currentRetryAttempts is declared for the sole purpose of code readability
+	isDestroyRun := true
+	currentRetryAttempts := 0
+	return createWorkspaceRun(d, meta, isDestroyRun, currentRetryAttempts)
+}
+
+func resourceTFEWorkspaceRunUpdate(d *schema.ResourceData, meta interface{}) error {
+	// update is a noop since this resource only creates a run during a destroy or an initial apply phase
+	return nil
+}
+
+func resourceTFEWorkspaceRunRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	log.Printf("[DEBUG] Read run for: %s", d.Id())
+	runID := d.Id()
+	_, err := config.Client.Runs.Read(ctx, runID)
+	if err != nil {
+		if errors.Is(err, tfe.ErrResourceNotFound) {
+			log.Printf("[DEBUG] Run %s does not exist", d.Id())
```

Comment:
The whole idea of a run as a resource is juuuust askew enough from Terraform's normal resource-centric view of the world that I'm not 100% positive that this standard "remove from state if 404 on read" logic makes sense. As long as we've got the early out I suggested above, I think it's probably fine, but we might as well make our logic crystal clear in the comments, since the correct behavior (or even how this situation could come about) is not at all obvious to a newcomer to this code.

```suggestion
			// It would be very strange for this to happen, since runs can't
			// normally be deleted independently. But this *probably* means we
			// never performed the initial apply, so we'll remove the missing
			// run from the state to force an apply to happen.
			log.Printf("[DEBUG] Run %s does not exist", d.Id())
```